### PR TITLE
LibJS: Hide all the constructors!

### PR DESCRIPTION
**Pull request by awesomekling**
Use the magic of friendship to remove the need for `JS::Cell` subclasses to have public constructors.

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Array.h [hunk 1]
```diff
@@ -45,6 +44,9 @@ class Array : public Object {
 
     [[nodiscard]] bool length_is_writable() const { return m_length_writable; };
 
+protected:
+    explicit Array(Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Array.h:48]: Why protected and not private for this one?

File: Userland/Libraries/LibJS/Runtime/ArrayIterator.h [hunk 2]
```diff
@@ -26,6 +25,8 @@ class ArrayIterator final : public Object {
 private:
     friend class ArrayIteratorPrototype;
 
+    explicit ArrayIterator(Value array, Object::PropertyKind iteration_kind, Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayIterator.h:28]: Could drop the explicit while we're here

File: Userland/Libraries/LibJS/Runtime/AsyncFromSyncIterator.h [hunk 3]
```diff
@@ -29,6 +28,8 @@ class AsyncFromSyncIterator final : public Object {
     Iterator const& sync_iterator_record() const { return m_sync_iterator_record; }
 
 private:
+    explicit AsyncFromSyncIterator(Realm&, Iterator sync_iterator_record);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncFromSyncIterator.h:31]: Likewise

File: Userland/Libraries/LibJS/Runtime/AsyncFunctionDriverWrapper.h [hunk 4]
```diff
@@ -19,14 +19,15 @@ class AsyncFunctionDriverWrapper final : public Promise {
...
     static ThrowCompletionOr<Value> create(Realm&, GeneratorObject*);
-    explicit AsyncFunctionDriverWrapper(Realm&, GeneratorObject*);
 
     virtual ~AsyncFunctionDriverWrapper() override = default;
     void visit_edges(Cell::Visitor&) override;
 
     ThrowCompletionOr<Value> react_to_async_task_completion(VM&, Value, bool is_successful);
 
 private:
+    explicit AsyncFunctionDriverWrapper(Realm&, GeneratorObject*);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/AsyncFunctionDriverWrapper.h:29]: Likewise

File: Userland/Libraries/LibJS/Runtime/DataView.h [hunk 5]
```diff
@@ -18,14 +18,15 @@ class DataView : public Object {
...
 
-    explicit DataView(ArrayBuffer*, size_t byte_length, size_t byte_offset, Object& prototype);
     virtual ~DataView() override = default;
 
     ArrayBuffer* viewed_array_buffer() const { return m_viewed_array_buffer; }
     size_t byte_length() const { return m_byte_length; }
     size_t byte_offset() const { return m_byte_offset; }
 
 private:
+    explicit DataView(ArrayBuffer*, size_t byte_length, size_t byte_offset, Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/DataView.h:28]: Likewise

File: Userland/Libraries/LibJS/Runtime/DataViewPrototype.h [hunk 6]
```diff
@@ -15,11 +15,12 @@ class DataViewPrototype final : public PrototypeObject<DataViewPrototype, DataVi
     JS_PROTOTYPE_OBJECT(DataViewPrototype, DataView, DataView);
 
 public:
-    DataViewPrototype(Realm&);
     virtual void initialize(Realm&) override;
     virtual ~DataViewPrototype() override = default;
 
 private:
+    DataViewPrototype(Realm&);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/DataViewPrototype.h:22]: ...but make this one explicit

File: Userland/Libraries/LibJS/Runtime/DeclarativeEnvironment.h [hunk 7]
```diff
@@ -62,6 +59,10 @@ class DeclarativeEnvironment : public Environment {
     ThrowCompletionOr<void> set_mutable_binding_direct(VM&, size_t index, Value, bool strict);
 
 protected:
+    DeclarativeEnvironment();
+    explicit DeclarativeEnvironment(Environment* parent_environment);
+    explicit DeclarativeEnvironment(Environment* parent_environment, Span<Binding const> bindings);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/DeclarativeEnvironment.h:64]: Likewise

File: Userland/Libraries/LibJS/Runtime/FinalizationRegistry.h [hunk 8]
```diff
@@ -38,6 +37,8 @@ class FinalizationRegistry final
     JobCallback const& cleanup_callback() const { return m_cleanup_callback; }
 
 private:
+    explicit FinalizationRegistry(Realm&, JobCallback, Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/FinalizationRegistry.h:40]: Likewise

File: Userland/Libraries/LibJS/Runtime/FinalizationRegistryPrototype.h [hunk 9]
```diff
@@ -15,11 +15,12 @@ class FinalizationRegistryPrototype final : public PrototypeObject<FinalizationR
     JS_PROTOTYPE_OBJECT(FinalizationRegistryPrototype, FinalizationRegistry, FinalizationRegistry);
 
 public:
-    FinalizationRegistryPrototype(Realm&);
     virtual void initialize(Realm&) override;
     virtual ~FinalizationRegistryPrototype() override = default;
 
 private:
+    FinalizationRegistryPrototype(Realm&);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/FinalizationRegistryPrototype.h:22]: Likewise

File: Userland/Libraries/LibJS/Runtime/Intl/DisplayNames.h [hunk 10]
```diff
@@ -63,6 +62,8 @@ class DisplayNames final : public Object {
     StringView language_display_string() const;
 
 private:
+    DisplayNames(Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/DisplayNames.h:65]: Likewise

File: Userland/Libraries/LibJS/Runtime/Intl/PluralRules.h [hunk 11]
```diff
@@ -19,14 +19,15 @@ class PluralRules final : public NumberFormatBase {
...
 public:
-    PluralRules(Object& prototype);
     virtual ~PluralRules() override = default;
 
     Unicode::PluralForm type() const { return m_type; }
     StringView type_string() const { return Unicode::plural_form_to_string(m_type); }
     void set_type(StringView type) { m_type = Unicode::plural_form_from_string(type); }
 
 private:
+    PluralRules(Object& prototype);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Intl/PluralRules.h:29]: Likewise

File: Userland/Libraries/LibJS/Runtime/Array.h (hunk 1 above)
Comment (awesomekling) [reply Userland/Libraries/LibJS/Runtime/Array.h:48]: ArrayPrototype inherits Array :acid2kiss:

**Review by linusg — APPROVED** [review]
LGTM!

**Pull request merged**